### PR TITLE
[eas-cli] create commands for managing environment secrets

### PR DESCRIPTION
**Pull request by FiberJW**
# Checklist

- [x] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary.

# Why

We want to have a CLI for managing encrypted environment secrets for hosted processes like EAS Build.

# How

- `secrets:list` lists out secrets associated with the current account and current app

Screen Shot 2021-03-15 at 17 42 41

- `secrets:create` creates a secret for either the current account or current app

Screen Shot 2021-03-15 at 17 46 47

Screen Shot 2021-03-15 at 17 46 15

- `secrets:delete` deletes a secret by ID

Screen Shot 2021-03-15 at 17 48 07

Screen Shot 2021-03-15 at 17 48 52

# Test Plan

`EXPO_STAGING=true easd secrets:<list | create | delete>` should work as illustrated by the screenshots above. The list of secrets you create should match what is represented in the web UI on https://staging.expo.io

### CODE REVIEW

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  isEasEnabledForProjectAsync,
+  warnEasUnavailable,
+} from '../../project/isEasEnabledForProject';
+import { findProjectRootAsync, getProjectAccountNameAsync } from '../../project/projectUtils';
+import { promptAsync } from '../../prompts';
+import { ensureLoggedInAsync } from '../../user/actions';
+
+export default class EnvironmentSecretCreate extends Command {
+  static hidden = true;
+  static description = 'Create an environment secret for on the current project or owner account.';
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:19]: Sth's wrong with this sentence.

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    if (!target) {
+      const validationMessage = 'Secret target may not be empty.';
+
+      ({ target } = await promptAsync({
+        type: 'select',
+        name: 'target',
+        message: 'Where should this secret be used:',
+        choices: [
+          { title: 'Account-wide', value: 'account' },
+          { title: 'Project-specific', value: 'project' },
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:73]: Maybe create an enum for that?
```
enum SecretTargetLocation {
  ACCOUNT = 'account',
  PROJECT = 'project',
}
```

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      }
+
+      Log.withTick(
+        `️Created a new secret ${chalk.bold(name)} on project ${chalk.bold(
+          `@${accountName}/${slug}`
+        )}.`
+      );
+    }
+
+    if (target === 'account') {
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:133]:
```suggestion
    } else if (target === 'account') {
```

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        name: 'secretValue',
+        message: 'Secret value:',
+        validate: value => (value ? true : validationMessage),
+      }));
+    }
+
+    let secret;
+
+    if (target === 'project') {
+      if (!projectDir) {
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:112]: I think that `projectDir` cannot be falsy, see https://github.com/expo/eas-cli/pull/275/files#diff-1a72c4b45ae1df867f274f0111db129f7c4cb44e8bacabb8fba129410eb1ba7fR46

File: packages/eas-cli/src/commands/secrets/delete.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    } = this.parse(EnvironmentSecretDelete);
+
+    Log.addNewLineIfNone();
+    Log.warn(
+      `You are about to permamently delete secret with id: "${id}".\nThis action is irreversable.`
+    );
+    Log.newLine();
+    const confirmed = await toggleConfirmAsync({ message: 'Are you sure you wish to proceed?' });
+    if (!confirmed) {
+      Log.error(`Cancelled deletion of secret with id: "${id}".`);
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/delete.ts:37]:
```suggestion
      Log.error(`Canceled deletion of secret with id: "${id}".`);
```
🇺🇸

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 6]
```diff
@@ -0,0 +1,153 @@
+import { getConfig } from '@expo/config';
+import { Command } from '@oclif/command';
+import chalk from 'chalk';
+
+import { AccountResolver } from '../../devices/manager';
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:5]: Please move the `AccountResolver` class to a more generic place.

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+      ({ secretValue } = await promptAsync({
+        type: 'text',
+        name: 'secretValue',
+        message: 'Secret value:',
+        validate: value => (value ? true : validationMessage),
+      }));
+    }
+
+    let secret;
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:109]:
```suggestion
```

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+    if (target === 'project') {
+      if (!projectDir) {
+        throw new Error('Please run this command inside a project directory.');
+      }
+
+      secret = await EnvironmentSecretMutation.createForApp(
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:116]:
```suggestion
      const secret = await EnvironmentSecretMutation.createForApp(
```

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+          `@${accountName}/${slug}`
+        )}.`
+      );
+    }
+
+    if (target === 'account') {
+      const resolver = new AccountResolver(projectDir, user);
+      const ownerAccount = await resolver.resolveAccountAsync();
+
+      secret = await EnvironmentSecretMutation.createForAccount(
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/create.ts:137]:
```suggestion
      const secret = await EnvironmentSecretMutation.createForAccount(
```

File: packages/eas-cli/src/commands/secrets/delete.ts [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    {
+      name: 'id',
+      required: true,
+      description: `ID of the secret to delete`,
+    },
+  ];
+
+  async run() {
+    await ensureLoggedInAsync();
+
```
Comment (dsokal) [inline packages/eas-cli/src/commands/secrets/delete.ts:25]: Maybe we should check here if EAS is enabled/available for the project? The same way you're checking it in the other two commands.

File: packages/eas-cli/src/graphql/mutations/EnvironmentSecretMutation.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  async createForAccount(
+    input: { name: string; value: string },
+    accountId: string
+  ): Promise<EnvironmentSecretFragment> {
+    const data = await withErrorHandlingAsync(
+      graphqlClient
+        .mutation<CreateEnvironmentSecretForAccountMutation>(
+          gql`
+            mutation CreateEnvironmentSecretForAccount(
+              $input: CreateEnvironmentSecretInput
```
Comment (dsokal) [inline packages/eas-cli/src/graphql/mutations/EnvironmentSecretMutation.ts:23]: Maybe we should make the input required in our API?
I can see it's not required https://github.com/expo/universe/blob/main/server/www/src/graphql/mutations/EnvironmentSecretMutation.ts#L27
but then it's treated as if it always exists - https://github.com/expo/universe/blob/main/server/www/src/graphql/mutations/EnvironmentSecretMutation.ts#L58

Comment (dsokal) [reply packages/eas-cli/src/graphql/mutations/EnvironmentSecretMutation.ts:23]: The same applies to `createEnvironmentSecretForApp`

**Review by dsokal — COMMENTED** [review]
Looks good!

**Comment by dsokal** [comment]
https://github.com/expo/eas-cli/pull/275/checks?check_run_id=2116645222 - see the linter error

File: packages/eas-cli/src/commands/secrets/create.ts [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      Log.withTick(
+        `️Created a new secret ${chalk.bold(name)} on project ${chalk.bold(
+          `@${accountName}/${slug}`
+        )}.`
+      );
+    }
+
+    if (target === 'account') {
+      const resolver = new AccountResolver(projectDir, user);
+      const ownerAccount = await resolver.resolveAccountAsync();
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/secrets/create.ts:135]: you already know `accountName` then you can just find is with findAccountByName from https://github.com/expo/eas-cli/blob/main/packages/eas-cli/src/user/Account.ts

File: packages/eas-cli/src/commands/secrets/list.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      EnvironmentSecretsQuery.byAppFullNameAsync(projectFullName),
+    ]);
+
+    const secrets = [
+      ...appSecrets.map(s => ({ ...s, type: 'app-specific' })),
+      ...accountSecrets.map(s => ({ ...s, type: 'account-wide' })),
+    ] as (EnvironmentSecretFragment & { type: 'app-specific' | 'account-wide' })[];
+
+    const table = new Table({
+      head: ['name', 'type', 'updated-at', 'id'],
```
Comment (barthap) [inline packages/eas-cli/src/commands/secrets/list.ts:53]: How about
```suggestion
      head: ['Name', 'Type', 'Updated at', 'ID'],
```

File: packages/eas-cli/src/commands/secrets/list.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      wordWrap: true,
+    });
+
+    for (const secret of secrets) {
+      const { name, createdAt: updatedAt, type, id } = secret;
+      table.push([name, type, updatedAt, id]);
```
Comment (barthap) [inline packages/eas-cli/src/commands/secrets/list.ts:59]: Maybe we can pretty-print the `updatedAt` datetime; I think that `2021-03-15T22:01:23.123Z` is a bit difficult to read. What about displaying it in `YYYY-MM-DD HH:mm:ss` format?

File: packages/eas-cli/src/graphql/mutations/EnvironmentSecretMutation.ts (hunk 11 above)
Comment (FiberJW) [reply packages/eas-cli/src/graphql/mutations/EnvironmentSecretMutation.ts:23]: ah yeah, thanks for noticing this! it was definitely intended to be required

**Comment by FiberJW** [comment]
Here's the updated list format:

Screen Shot 2021-03-16 at 16 35 51

File: packages/eas-cli/src/devices/manager.ts [hunk 15]
```diff
@@ -44,7 +44,7 @@ export default class DeviceManager {
   }
 }
 
-export class AccountResolver {
+class AccountResolver {
   constructor(private projectDir: string | null, private user: Actor) {}
 
   public async resolveAccountAsync(): Promise<Account> {
```
Comment (dsokal) [inline packages/eas-cli/src/devices/manager.ts:50]: Revert

**Review by dsokal — APPROVED** [review]
🚀  Fix tests before merging.

**Review by barthap — APPROVED** [review]
LGTM 👍 Such a lovely CRUD 😁 Just tests need to be fixed.

BTW, I think I have an idea for future improvement of `eas secrets:delete` - maybe we can display an interactive list from which users may choose secret to delete (the same way we select Apple Team or Distribution Certificates). This would allow users to skip copy-pasting secret ID from `eas secrets:list` command 😉 Anyway, for now current implementation seems sufficient.

**Pull request merged**